File: LaRC/CAL_LID_L2_PSCMask-Standard-V2-00.2021-03-26T00-00-00ZN.hdf.v.py
"""
This example code illustrates how to access and visualize a LaRC CALIPSO
LIDAR L2 PSCMask HDF4 file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python CAL_LID_L2_PSCMask-Standard-V2-00.2021-03-26T00-00-00ZN.hdf.v.py

The HDF file must be in your current working directory.

Tested under: Python 3.9.12 :: Miniconda
Last updated: 2022-09-23
"""
import os
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt

from pyhdf.SD import SD, SDC
from matplotlib import colors
from mpl_toolkits.basemap import Basemap

FILE_NAME = 'CAL_LID_L2_PSCMask-Standard-V2-00.2021-03-26T00-00-00ZN.hdf'

# Read data field.
DATAFIELD_NAME = 'PSC_Feature_Mask'

hdf = SD(FILE_NAME, SDC.READ)
        
# Read dataset.
data2D = hdf.select(DATAFIELD_NAME)
data = data2D[:,:].astype(np.float64)

# Retrieve the attributes.
attrs = data2D.attributes(full=1)
# print(attrs)

fva=attrs["fillvalue "]
fill_value = fva[0]
ua=attrs["units "]
units = ua[0]

 # Replace the missing values with NaN.        
data[data == float(fill_value)] = np.nan
datam = np.ma.masked_array(data, np.isnan(data))
    
# Read geolocation datasets.
latitude = hdf.select('Latitude')
lat = latitude[:]
        
altitude = hdf.select('Altitude')
alt = altitude[:]

# Subset latitude values that decrease monotonically.
s = 0
e = 566

lat = lat[s:e]
size = lat.shape[0]

datam = datam[s:e, :]
      
# Contour the data on a grid of latitude vs. altitude
latitude, altitude = np.meshgrid(lat, alt)

basename = os.path.basename(FILE_NAME)
plt.contourf(latitude, altitude, datam.T)
cb = plt.colorbar()
cb.set_label('Unit:'+units)
plt.title('{0}\n{1}'.format(basename, DATAFIELD_NAME))
plt.xlabel('Latitude (degrees north)')
plt.ylabel('Altitude (km)')
fig = plt.gcf()

pngfile = "{0}.v.py.png".format(basename)
fig.savefig(pngfile)
    

 

File: LaRC/CAL_LID_L2_PSCMask_Standard_V2_00_2021_03_26T00_00_00ZN_hdf_v.m
%  This example code illustrates how to access and visualize LaRC CALIPSO LIDAR
% Level 2 PCSMask HDF4 file in MATLAB. 
%
%  If you have any questions, suggestions, comments on this example, 
% please use the HDF-EOS Forum (http://hdfeos.org/forums). 
% 
%  If you would like to see an  example of any other NASA HDF/HDF-EOS
% data product that is not listed in the HDF-EOS  Comprehensive Examples 
% page (http://hdfeos.org/zoo), feel free to contact us at 
% eoshelp@hdfgroup.org or post it at the HDF-EOS Forum 
% (http://hdfeos.org/forums).
%
% Usage:save this script and run (without .m at the end)
%
%  $matlab -nosplash -nodesktop -r CAL_LID_L2_PSCMask_Standard_V2_00_2021_03_26T00_00_00ZN_hdf_v
%
% Tested under: MATLAB R2021a
% Last updated: 2022-09-26

import matlab.io.hdf4.*

% Open the HDF4 File.
FILE_NAME = 'CAL_LID_L2_PSCMask-Standard-V2-00.2021-03-26T00-00-00ZN.hdf';
SD_id = sd.start(FILE_NAME, 'rdonly');

% Read data.
datafield_name='PSC_Feature_Mask';
sds_index = sd.nameToIndex(SD_id, datafield_name);
sds_id = sd.select(SD_id, sds_index);
data = sd.readData(sds_id);
sd.endAccess(sds_id);

% Read _FillValue from data field.
fillvalue_index = sd.findAttr(sds_id, 'fillvalue ');
fillvalue = str2double(sd.readAttr(sds_id, fillvalue_index));

% Read units attribute from data field.
units_index = sd.findAttr(sds_id, 'units ');
units = sd.readAttr(sds_id, units_index);

% Read lat.
lat_name='Latitude';
sds_index = sd.nameToIndex(SD_id, lat_name);
sds_id = sd.select(SD_id, sds_index);
lat = sd.readData(sds_id);
sd.endAccess(sds_id);

% Read altitude.
alt_name='Altitude';
sds_index = sd.nameToIndex(SD_id, alt_name);
sds_id = sd.select(SD_id, sds_index);
alt = sd.readData(sds_id);
sd.endAccess(sds_id);

% Close the file.
sd.close(SD_id);

% Convert data to double type for plot.
data=double(data);
alt=double(alt);
lat=double(lat);

% Replace the fill value with NaN.
data(data==fillvalue) = NaN;

% Subset latitude values that decrease monotonically.
s = 1;
e = 567;
lat = lat(s:e);
data = data(:,s:e);

% Plot over altitude = Y-axis / latitude = X-axis.
% data = data';


% Create a Figure to Plot the data.
f = figure('Name', FILE_NAME, ...
    'Renderer', 'zbuffer', ...
    'Position', [0,0,800,600], ...
    'Visible', 'off', ...
    'PaperPositionMode', 'auto');
colormap('Jet');

% contourf(lat, alt, rot90(data, 1));
contourf(lat, alt, data);

% Set axis labels.
xlabel('Latitude (degrees north)'); 
ylabel('Altitude (km)');

% Put colorbar.
h = colorbar();

tstring = {FILE_NAME;[datafield_name, ' (', units, ')']};
title(tstring, 'Interpreter', 'none', 'FontSize', 16, ...
      'FontWeight','bold');

hold off;
saveas(f, [FILE_NAME '.v.m.png']);
exit;

